src/cycles/eulerian.jl

```diff
@@ -0,0 +1,318 @@
+# Checking for existence of Eulerian circuit and Eulerian trail
+
+function single_nonzero_degree_component(
+    graph::AG,
+    ) where {T, AG <: AbstractGraph{T}}
+
+    visited = zeros(UInt8, nv(graph))
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            visited[v] = 2
+        end
+    end
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            continue
+        end
+        S = Vector{T}([v])
+        visited[v] = 1
+        while !isempty(S)
+            u = S[end]
+            w = 0
```

Comment:
Use  `w = zero(eltype(graph))` otherwise the type of `w` will change when you assign `n` to it later, as `n` might not be of type `Int`. This leads to type instability and worse performance.

Reply:
Also maybe add the comment, that you are assuming that valid vertices never take the value `0`.